http4s/server/src/main/scala/endpoints/http4s/server/Urls.scala

```diff
@@ -0,0 +1,196 @@
+      paths match {
+        case head :: tail if head == segment => Some(Right(((), tail)))
+        case _                               => None
+      }
+  }
+
+  def segment[A](name: String = "", docs: Documentation = None)(
+      implicit A: Segment[A]): Path[A] = new Path[A] {
+    def decode(segments: List[String]): Option[Either[ErrorResponse, (A, List[String])]] = {
+      def uncons[B](bs: List[B]): Option[(B, List[B])] =
+        bs match {
+          case head :: tail => Some((head, tail))
+          case Nil          => None
+        }
+
+      uncons(segments).map {
+        case (s, ss) =>
+          A(s) match {
+            case None    => Left(badRequestResponse)
+            case Some(a) => Right((a, ss))
+          }
+      }
+    }
+  }
+
+  implicit def stringSegment: Segment[String] = Some(_)
+
+  def remainingSegments(name: String = "",
+                        docs: Documentation = None): Path[String] =
+    new Path[String] {
+      def decode(segments: List[String]): Option[Either[ErrorResponse, (String, List[String])]] =
+        if (segments.isEmpty) None
+        else Some(Right((segments.map(URLEncoder.encode(_, utf8Name)).mkString("/"), Nil)))
+    }
+
+  def chainPaths[A, B](first: Path[A], second: Path[B])(
+      implicit tupler: Tupler[A, B]): Path[tupler.Out] = new Path[tupler.Out] {
+    override def decode(
+        segments: List[String]): Option[Either[ErrorResponse, (tupler.Out, List[String])]] =
+      first.decode(segments).flatMap {
+        case Right((a, segments2)) =>
+          second.decode(segments2).map {
+            case Right((b, segments3)) => Right((tupler(a, b), segments3))
+            case Left(error)           => Left(error)
+          }
+        case Left(error) => Some(Left(error))
+      }
+  }
+
+  implicit def urlPartialInvFunctor: PartialInvariantFunctor[Url] =
+    new PartialInvariantFunctor[Url] {
+      override def xmapPartial[A, B](fa: Url[A],
+                                     f: A => Option[B],
+                                     g: B => A): Url[B] = new Url[B] {
+        override def decodeUrl(uri: Uri): Option[Either[ErrorResponse, B]] =
+          fa.decodeUrl(uri).map(_.right.flatMap(a => f(a).toRight(badRequestResponse)))
+      }
+    }
+
+  /** Builds an URL from the given path and query string */
+  def urlWithQueryString[A, B](path: Path[A], qs: QueryString[B])(
+      implicit tupler: Tupler[A, B]): Url[tupler.Out] = new Url[tupler.Out] {
+    override def decodeUrl(uri: Uri): Option[Either[ErrorResponse, tupler.Out]] = {
+      pathExtractor(path, uri).map(_.right.flatMap { a =>
+        qs(uri.multiParams) match {
+          case None    => Left(badRequestResponse)
+          case Some(b) => Right(tupler(a, b))
+        }
+      })
+    }
+  }
+
+  private def pathExtractor[A](path: Path[A], uri: http4s.Uri): Option[Either[ErrorResponse, A]] = {
+    val segments =
+      uri.path
+        .split("/")
+      .map(URLDecoder.decode(_, utf8Name))
```

Comment:
We definietely need to reconsider scalafmt for everyone benefit. @julienrf Are you sitll against it ?

Reply:
We can try, OK!